### PR TITLE
MEN-4745 move the service to one db for all tenants

### PR DESCRIPTION
**Pull request by MaciejTe**
refactor: move the service to the one database

Changelog: title

Signed-off-by: Maciej Tomczuk <maciej.tomczuk@northern.tech>

### CODE REVIEW

File: store/mongo/migration_2_0_0.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+			Indexes: []mongo.IndexModel{
+				{
+					Keys: bson.D{{Key: "exp.time", Value: 1}},
+					Options: &mopts.IndexOptions{
+						ExpireAfterSeconds: &expireAfterSeconds,
+						Name:               &tokenExpirationIndexName,
+					},
+				},
+			},
+		},
```
Comment (alfrunes) [inline store/mongo/migration_2_0_0.go:78]: I think we can leave the tokens collection as is. Although this will lead to all devices to become unauthorized on the first request after the migration. Otherwise, I think this collection needs to be treated separately.

File: store/mongo/migration_2_0_0.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	collectionsIndexes := map[string]struct {
+		Indexes []mongo.IndexModel
+	}{
+		DbDevicesColl: {
+			Indexes: []mongo.IndexModel{
+				// create device index on status and id
+				{
+					Keys: bson.D{
+						{Key: model.DevKeyId, Value: 1},
+						{Key: model.TenantIDField, Value: 1},
```
Comment (alfrunes) [inline store/mongo/migration_2_0_0.go:49]: I think the TenantIDField should be the top-level key.

File: store/mongo/migration_2_0_0.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+						{Key: model.AuthSetKeyDeviceId, Value: 1},
+						{Key: model.AuthSetKeyIdDataSha256, Value: 1},
+						{Key: model.AuthSetKeyPubKey, Value: 1},
+					},
+					Options: &options.IndexOptions{
+						Background: &_false,
+						Name:       &indexAuthSet_DeviceId_IdentityDataSha256_PubKey,
+						Unique:     &_true,
+					},
+				},
```
Comment (alfrunes) [inline store/mongo/migration_2_0_0.go:103]: Uhh... Isn't this index completely redundant? The above index is unique and contains a subset of the keys...

File: store/mongo/migration_2_0_0.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+				{
+					Keys: bson.D{
+						{Key: model.DevKeyId, Value: 1},
+						{Key: model.TenantIDField, Value: 1},
+					},
+					Options: &mopts.IndexOptions{
+						Background: &_false,
+						Name:       &indexDevices_Status,
+						Unique:     &_false,
+					},
```
Comment (alfrunes) [inline store/mongo/migration_2_0_0.go:55]: The options have a cleaner functional interface :)
Also, the `Background` option is deprecated since mongo 4.2.
```suggestion
					Options: mopts.Index().
						SetName(indexDevices_Status).
						SetUnique(false),
```

File: store/mongo/migration_2_0_0.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+						Name:               &tokenExpirationIndexName,
+					},
+				},
+			},
+		},
+		DbAuthSetColl: {
+			Indexes: []mongo.IndexModel{
+				{
+					Keys: bson.D{
+						{Key: model.AuthSetKeyIdDataSha256, Value: 1},
```
Comment (alfrunes) [inline store/mongo/migration_2_0_0.go:83]: I think we should add the `tenant_id` as the top level key in case of collisions.

File: cmd/commands.go [hunk 6]
```diff
@@ -463,95 +461,3 @@ func tryPropagateIdDataInventoryForDb(
...
-			}
-		}
-		return nil
-	}
-	// Start looping through the databases.
-	return ds.ForEachDatabase(
-		context.Background(),
-		mapFunc,
-	)
-}
```
Comment (alfrunes) [inline cmd/commands.go:557]: I think this deserves a separate commit.

File: cmd/commands_test.go [hunk 7]
```diff
@@ -393,248 +388,3 @@ func TestPropagateStatusesInventory(t *testing.T) {
 		})
 	}
 }
-
-func TestCheckDeviceLimits(t *testing.T) {
```
Comment (alfrunes) [inline cmd/commands_test.go:397]: ^

File: store/mongo/datastore_mongo.go [hunk 8]
```diff
@@ -114,58 +113,6 @@ func NewDataStoreMongo(config DataStoreMongoConfig) (*DataStoreMongo, error) {
 	return NewDataStoreMongoWithClient(client), nil
 }
 
-func (db *DataStoreMongo) ForEachDatabase(
```
Comment (alfrunes) [inline store/mongo/datastore_mongo.go:118]: ^

File: store/mongo/datastore_mongo.go [hunk 9]
```diff
@@ -224,7 +171,7 @@ func (db *DataStoreMongo) GetDevices(
 		findOpts.SetLimit(int64(limit))
 	}
 
-	cursor, err := collDevs.Find(ctx, fltr, findOpts)
+	cursor, err := collDevs.Find(ctx, ctxstore.WithTenantID(ctx, fltr), findOpts)
```
Comment (alfrunes) [inline store/mongo/datastore_mongo.go:174]: I think this belongs to the parent commit.

**Review by alfrunes — CHANGES_REQUESTED** [review]
The commit structure is a little off. Also remember we're using semantic commit messages now ;)
I also added some suggestions to the migrations.
There seems to be some issues with the unit- and acceptance tests still.

File: store/mongo/datastore_mongo.go [hunk 10]
```diff
@@ -432,126 +431,6 @@ func (db *DataStoreMongo) DeleteTokenByDevId(
...
-	}
-
-	err = m.Apply(ctx, *ver, migrations)
-	if err != nil {
-		return errors.Wrap(err, "failed to apply migrations")
-	}
-
-	return nil
-}
-
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply store/mongo/datastore_mongo.go:554]: Done!

File: store/mongo/datastore_mongo_test.go [hunk 11]
```diff
@@ -847,207 +840,6 @@ func verifyIndexes(t *testing.T, coll *mongo.Collection, expected []mongo.IndexM
...
-				assert.Equal(t, *v, out[len(out)-1].Version)
-
-				// verify that all indexes are created
-			} else {
-				assert.EqualError(t, err, tc.err)
-			}
-		})
-	}
-}
-
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply store/mongo/datastore_mongo_test.go:1050]: done!

File: cmd/commands.go (hunk 6 above)
Comment (MaciejTe) [reply cmd/commands.go:557]: Got it!

File: store/mongo/datastore_mongo.go (hunk 9 above)
Comment (MaciejTe) [reply store/mongo/datastore_mongo.go:174]: Done!

File: store/mongo/datastore_mongo.go (hunk 8 above)
Comment (MaciejTe) [reply store/mongo/datastore_mongo.go:118]: Done!

File: cmd/commands_test.go (hunk 7 above)
Comment (MaciejTe) [reply cmd/commands_test.go:397]: Done!

File: store/mongo/migration_2_0_0.go (hunk 5 above)
Comment (MaciejTe) [reply store/mongo/migration_2_0_0.go:79]: Done!

File: store/mongo/migration_2_0_0.go (hunk 3 above)
Comment (MaciejTe) [reply store/mongo/migration_2_0_0.go]: Done!

File: store/mongo/migration_2_0_0.go (hunk 4 above)
Comment (MaciejTe) [reply store/mongo/migration_2_0_0.go]: Thanks! Refactored

File: store/mongo/migration_2_0_0.go (hunk 2 above)
Comment (MaciejTe) [reply store/mongo/migration_2_0_0.go]: done!

File: store/mongo/migration_2_0_0.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+	ds  *DataStoreMongo
+	ctx context.Context
+}
+
+// Up creates an index on status and id in the devices collection
+func (m *migration_2_0_0) Up(from migrate.Version) error {
+	// _false := false
+	// _true := true
+	// var expireAfterSeconds int32 = 0
+	// tokenExpirationIndexName := "TokenExpiration"
```
Comment (alfrunes) [inline store/mongo/migration_2_0_0.go:45]:
```suggestion
```

File: store/mongo/migration_2_0_0.go [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+						{Key: model.TenantIDField, Value: 1},
+						{Key: model.DevKeyStatus, Value: 1},
+						{Key: model.DevKeyId, Value: 1},
+					},
+					Options: mopts.Index().
+						SetName(indexDevices_Status).
+						SetUnique(false),
+				}, {
+					Keys: bson.D{
+						{Key: model.DevKeyIdDataSha256, Value: 1},
```
Comment (alfrunes) [inline store/mongo/migration_2_0_0.go:66]: We need the `tenant_id` top-level key here as well to protect against cross-tenant collisions.

File: store/mongo/migration_2_0_0.go [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+						{Key: model.TenantIDField, Value: 1},
+						{Key: model.AuthSetKeyIdDataSha256, Value: 1},
+						{Key: model.AuthSetKeyPubKey, Value: 1},
+					},
+					Options: mopts.Index().
+						SetName(indexAuthSet_IdentityDataSha256_PubKey).
+						SetUnique(true),
+				},
+			},
+		},
```
Comment (alfrunes) [inline store/mongo/migration_2_0_0.go:87]: Reviewing the queries we make to the `auth_sets` collection, we'll also need an index key on the `device_id` field too. That is,
```json
{"tenant_id": 1, "device_id": 1}
```
You may also consider appending the a key on the timestamp as well to get the ability to sort the autsets for a device by date.

File: store/mongo/migration_2_0_0.go [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+				}
+			}
+			if len(writes) > 0 {
+				_, err := collOut.BulkWrite(m.ctx, writes)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
```
Comment (alfrunes) [inline store/mongo/migration_2_0_0.go:162]: We also need to handle the tokens collection (but separately). We need to make sure that the TTL index exists for the `exp` token claim. See `migration_1_8_0.go`.

File: store/mongo/datastore_mongo.go [hunk 16]
```diff
@@ -303,12 +300,12 @@ func (db *DataStoreMongo) AddToken(ctx context.Context, t *jwt.Token) error {
...
 
 	filter := bson.M{"_id": t.Claims.ID}
-	update := bson.M{"$set": t}
+	update := bson.M{"$set": ctxstore.WithTenantID(ctx, t)}
 	updateOpts := mopts.Update()
 	updateOpts.SetUpsert(true)
 
 	if _, err := collTokens.UpdateOne(
-		ctx, filter, update, updateOpts,
+		ctx, ctxstore.WithTenantID(ctx, filter), update, updateOpts,
```
Comment (alfrunes) [inline store/mongo/datastore_mongo.go:308]: Token claims should not have another tenant_id field. (It's already present in `mender.tenant` field).
```suggestion
		ctx, filter, update, updateOpts,
```

File: store/mongo/datastore_mongo.go [hunk 17]
```diff
@@ -342,7 +338,7 @@ func (db *DataStoreMongo) DeleteToken(ctx context.Context, jti oid.ObjectID) err
 	collTokens := database.Collection(DbTokensColl)
 
 	filter := bson.M{"_id": jti}
-	result, err := collTokens.DeleteOne(ctx, filter)
+	result, err := collTokens.DeleteOne(ctx, ctxstore.WithTenantID(ctx, filter))
```
Comment (alfrunes) [inline store/mongo/datastore_mongo.go:341]:
```suggestion
	result, err := collTokens.DeleteOne(ctx, filter)
```

File: store/mongo/datastore_mongo.go [hunk 18]
```diff
@@ -366,7 +362,7 @@ func (db *DataStoreMongo) DeleteTokenByDevId(
 ) error {
 	c := db.client.Database(ctxstore.DbFromContext(ctx, DbName)).
 		Collection(DbTokensColl)
-	ci, err := c.DeleteMany(ctx, bson.M{"sub": devID})
+	ci, err := c.DeleteMany(ctx, ctxstore.WithTenantID(ctx, bson.M{"sub": devID}))
```
Comment (alfrunes) [inline store/mongo/datastore_mongo.go:365]:
```suggestion
	var tenantID string
	if id := identity.FromContext(ctx); id != nil {
		tenantID = id.Tenant
	}
	ci, err := c.DeleteMany(ctx, bson.M{"sub": devID, "mender.tenant": tenantID})
```

Comment (alfrunes) [reply store/mongo/datastore_mongo.go:365]: By the way, there's no index backing this query. We should analyze the usage of this function (NOTE to self).

**Review by alfrunes — CHANGES_REQUESTED** [review]
This is as far as I got today. We need one more index in the `auth_sets` collection and at least keep the TTL index for the tokens collection.

File: store/mongo/datastore_mongo.go (hunk 16 above)
Comment (MaciejTe) [reply store/mongo/datastore_mongo.go]: `mender.tenant` wouldn't work because of struct `inline` and `omitempty` bson tags. Currently `tenant` field is used to filter tokens.

File: store/mongo/migration_2_0_0.go (hunk 15 above)
Comment (MaciejTe) [reply store/mongo/migration_2_0_0.go:183]: Done!

File: store/mongo/datastore_mongo.go (hunk 17 above)
Comment (MaciejTe) [reply store/mongo/datastore_mongo.go]: see above comment

File: store/mongo/migration_2_0_0.go (hunk 13 above)
Comment (MaciejTe) [reply store/mongo/migration_2_0_0.go:67]: Done!

**Comment by MaciejTe** [comment]
@mender-test-bot start pipeline

**Comment by MaciejTe** [comment]
@mender-test-bot start pipeline

**Review by kjaskiewiczz — CHANGES_REQUESTED** [review]
AFAIK device limits will not work - we're using limit name as an _id, so, without introducing some changes here, it will be impossible to set different limits for different tenants
